cabal-validate/src/Cli.hs

```diff
@@ -0,0 +1,433 @@
+              , optional (rawRunCliSuite opts && not (rawLibOnly opts)) CliSuite
+              , optionals (rawSolverBenchmarks opts) [SolverBenchmarksTests, SolverBenchmarksRun]
+              , [TimeSummary]
+              ]
+
+      targets' =
+        concat
+          [
+            [ "Cabal"
+            , "Cabal-hooks"
+            , "cabal-testsuite"
+            , "Cabal-tests"
+            , "Cabal-QuickCheck"
+            , "Cabal-tree-diff"
+            , "Cabal-described"
+            ]
+          , optionals
+              (CliTests `elem` steps')
+              [ "cabal-install"
+              , "cabal-install-solver"
+              , "cabal-benchmarks"
+              ]
+          , optional (rawSolverBenchmarks opts) "solver-benchmarks"
+          ]
+
+      archPath' =
+        let osPath =
+              case os of
+                "darwin" -> "osx"
+                "linux" -> "linux"
+                "mingw32" -> "windows"
+                _ -> os -- TODO: Warning?
+         in arch <> "-" <> osPath
+
+      projectFile' =
+        if rawLibOnly opts
+          then "cabal.validate-libonly.project"
+          else "cabal.validate.project"
+
+      tastyArgs' =
+        "--hide-successes"
+          : case rawTastyPattern opts of
+            Just tastyPattern -> ["--pattern", tastyPattern]
+            Nothing -> []
+
+  when (rawListSteps opts) $ do
+    -- TODO: This should probably list _all_ available steps, not just the selected ones!
+    putStrLn "Targets:"
+    forM_ targets' $ \target -> do
+      putStrLn $ "  " <> target
+    putStrLn "Steps:"
+    forM_ steps' $ \step -> do
+      putStrLn $ "  " <> displayStep step
+    exitSuccess
+
+  startTime' <- getAbsoluteTime
+  jobs' <- maybe getNumCapabilities pure (rawJobs opts)
+  cwd' <- getCurrentDirectory
+  compiler' <- makeCompiler (rawCompiler opts)
+
+  pure
+    Opts
+      { verbose = rawVerbose opts
+      , jobs = jobs'
+      , cwd = cwd'
+      , startTime = startTime'
+      , compiler = compiler'
+      , extraCompilers = rawExtraCompilers opts
+      , cabal = rawCabal opts
+      , archPath = archPath'
+      , projectFile = projectFile'
+      , hackageTests = rawHackageTests opts
+      , tastyArgs = tastyArgs'
+      , targets = targets'
+      , steps = steps'
+      }
+
+-- | Raw command-line options.
+data RawOpts = RawOpts
+  { rawVerbose :: Bool
+  , rawJobs :: Maybe Int
+  , rawCompiler :: FilePath
+  , rawCabal :: FilePath
+  , rawExtraCompilers :: [FilePath]
+  , rawTastyPattern :: Maybe String
+  , rawDoctest :: Bool
+  , rawSteps :: [Step]
+  , rawListSteps :: Bool
+  , rawLibOnly :: Bool
+  , rawRunLibTests :: Bool
+  , rawRunCliTests :: Bool
+  , rawRunLibSuite :: Bool
+  , rawRunCliSuite :: Bool
+  , rawSolverBenchmarks :: Bool
+  , rawHackageTests :: HackageTests
+  }
+  deriving (Show)
+
+-- | `Parser` for `RawOpts`.
+--
+-- See: `fullOptsParser`
+optsParser :: Parser RawOpts
```

Comment:
I’d expect this be named `rawOptsParser` (likewise for the “full” version below).